From f0462934f42eb692f416fe351b10fb40cdc7bfae Mon Sep 17 00:00:00 2001
From: Nitesh Gupta <nitegupt@qti.qualcomm.com>
Date: Wed, 21 May 2025 14:25:35 +0530
Subject: [PATCH 1/1] PCI:qcom:ep Add support for SCMI based PCIe EP for
 sa8797p

Add support for sa8797p for SCMI based Endpoint mode.

Change-Id: If7549a51d1362a1f6b8408f365ba4621be7145de
Signed-off-by: Nitesh Gupta <nitegupt@qti.qualcomm.com>
---
 drivers/pci/controller/dwc/pcie-qcom-ep.c    | 84 +++++++++++++++-----
 drivers/pci/endpoint/functions/pci-epf-mhi.c | 18 +++++
 2 files changed, 82 insertions(+), 20 deletions(-)

diff --git a/drivers/pci/controller/dwc/pcie-qcom-ep.c b/drivers/pci/controller/dwc/pcie-qcom-ep.c
index 72bbac1d4e23..d5e7b1464179 100644
--- a/drivers/pci/controller/dwc/pcie-qcom-ep.c
+++ b/drivers/pci/controller/dwc/pcie-qcom-ep.c
@@ -24,6 +24,7 @@
 #include <linux/regmap.h>
 #include <linux/reset.h>
 #include <linux/module.h>
+#include <linux/iopoll.h>
 
 #include "../../pci.h"
 #include "pcie-designware.h"
@@ -158,6 +159,13 @@
 
 #define to_pcie_ep(x)				dev_get_drvdata((x)->dev)
 
+#define PCS_COM_SW_RESET			0x60
+#define PCS_COM_PCS_STATUS1			0x10
+#define PCS_COM_POWER_DOWN_CONTROL		0x64
+#define PCS_COM_START_CONTROL			0x5C
+
+#define PHY_INIT_COMPLETE_TIMEOUT		10000
+#define PHY_STATUS				BIT(7)
 enum qcom_pcie_ep_link_status {
 	QCOM_PCIE_EP_LINK_DISABLED,
 	QCOM_PCIE_EP_LINK_ENABLED,
@@ -177,6 +185,7 @@ struct qcom_pcie_ep_cfg {
 	u32	hdma_wr_channels;
 	bool	override_no_snoop;
 	bool	disable_mhi_ram_parity_check;
+	bool scmi_support;
 };
 
 /**
@@ -385,6 +394,8 @@ static int qcom_pcie_enable_resources(struct qcom_pcie_ep *pcie_ep)
 
 static void qcom_pcie_disable_resources(struct qcom_pcie_ep *pcie_ep)
 {
+	if (pcie_ep->cfg && pcie_ep->cfg->scmi_support)
+		return;
 	icc_set_bw(pcie_ep->icc_mem, 0, 0);
 	phy_power_off(pcie_ep->phy);
 	phy_exit(pcie_ep->phy);
@@ -395,13 +406,38 @@ static int qcom_pcie_perst_deassert(struct dw_pcie *pci)
 {
 	struct qcom_pcie_ep *pcie_ep = to_pcie_ep(pci);
 	struct device *dev = pci->dev;
+	struct platform_device *pdev = to_platform_device(dev);
+	void __iomem *status;
+	void __iomem *phy_pcs;
+	struct resource *res;
 	u32 val, offset;
 	int ret;
 
-	ret = qcom_pcie_enable_resources(pcie_ep);
-	if (ret) {
-		dev_err(dev, "Failed to enable resources: %d\n", ret);
-		return ret;
+	if (pcie_ep->cfg && pcie_ep->cfg->scmi_support) {
+		res = platform_get_resource_byname(pdev, IORESOURCE_MEM, "phy_pcs");
+		phy_pcs = ioremap(res->start, resource_size(res));
+
+		writel_relaxed(0x1, phy_pcs + PCS_COM_POWER_DOWN_CONTROL);
+		/* Pull PHY out of reset state */
+		writel_relaxed(0x0, phy_pcs + PCS_COM_SW_RESET);
+		/* start SerDes and Phy-Coding-Sublayer */
+		writel_relaxed(0x3, phy_pcs + PCS_COM_START_CONTROL);
+		usleep_range(1000, 1200);
+		status = phy_pcs + PCS_COM_PCS_STATUS1;
+
+		ret = readl_poll_timeout(status, val, !(val & PHY_STATUS), 200,
+					 PHY_INIT_COMPLETE_TIMEOUT);
+		if (ret) {
+			dev_err(pci->dev, "Phy initialization timed-out\n");
+			goto phy_power_off;
+		}
+	}
+	else {
+		ret = qcom_pcie_enable_resources(pcie_ep);
+		if (ret) {
+			dev_err(dev, "Failed to enable resources: %d\n", ret);
+			return ret;
+		}
 	}
 
 	/* Assert WAKE# to RC to indicate device is ready */
@@ -542,6 +578,10 @@ static int qcom_pcie_perst_deassert(struct dw_pcie *pci)
 
 	return 0;
 
+phy_power_off:
+	writel_relaxed(0x0, phy_pcs + PCS_COM_START_CONTROL);
+	writel_relaxed(0x0, phy_pcs + PCS_COM_POWER_DOWN_CONTROL);
+
 err_disable_resources:
 	qcom_pcie_disable_resources(pcie_ep);
 
@@ -640,16 +680,6 @@ static int qcom_pcie_ep_get_resources(struct platform_device *pdev,
 		return ret;
 	}
 
-	pcie_ep->num_clks = devm_clk_bulk_get_all(dev, &pcie_ep->clks);
-	if (pcie_ep->num_clks < 0) {
-		dev_err(dev, "Failed to get clocks\n");
-		return pcie_ep->num_clks;
-	}
-
-	pcie_ep->core_reset = devm_reset_control_get_exclusive(dev, "core");
-	if (IS_ERR(pcie_ep->core_reset))
-		return PTR_ERR(pcie_ep->core_reset);
-
 	pcie_ep->reset = devm_gpiod_get(dev, "reset", GPIOD_IN);
 	if (IS_ERR(pcie_ep->reset))
 		return PTR_ERR(pcie_ep->reset);
@@ -658,13 +688,25 @@ static int qcom_pcie_ep_get_resources(struct platform_device *pdev,
 	if (IS_ERR(pcie_ep->wake))
 		return PTR_ERR(pcie_ep->wake);
 
-	pcie_ep->phy = devm_phy_optional_get(dev, "pciephy");
-	if (IS_ERR(pcie_ep->phy))
-		ret = PTR_ERR(pcie_ep->phy);
+	if (!(pcie_ep->cfg && pcie_ep->cfg->scmi_support)) {
+		pcie_ep->num_clks = devm_clk_bulk_get_all(dev, &pcie_ep->clks);
+		if (pcie_ep->num_clks < 0) {
+			dev_err(dev, "Failed to get clocks\n");
+			return pcie_ep->num_clks;
+		}
+
+		pcie_ep->core_reset = devm_reset_control_get_exclusive(dev, "core");
+		if (IS_ERR(pcie_ep->core_reset))
+			return PTR_ERR(pcie_ep->core_reset);
 
-	pcie_ep->icc_mem = devm_of_icc_get(dev, "pcie-mem");
-	if (IS_ERR(pcie_ep->icc_mem))
-		ret = PTR_ERR(pcie_ep->icc_mem);
+		pcie_ep->phy = devm_phy_optional_get(dev, "pciephy");
+		if (IS_ERR(pcie_ep->phy))
+			ret = PTR_ERR(pcie_ep->phy);
+
+		pcie_ep->icc_mem = devm_of_icc_get(dev, "pcie-mem");
+		if (IS_ERR(pcie_ep->icc_mem))
+			ret = PTR_ERR(pcie_ep->icc_mem);
+	}
 
 	return ret;
 }
@@ -935,9 +977,11 @@ static const struct qcom_pcie_ep_cfg cfg_1_34_0 = {
 	.hdma_wr_channels = 1,
 	.override_no_snoop = true,
 	.disable_mhi_ram_parity_check = true,
+	.scmi_support = true,
 };
 
 static const struct of_device_id qcom_pcie_ep_match[] = {
+	{ .compatible = "qcom,sa8797p-pcie-ep", .data = &cfg_1_34_0},
 	{ .compatible = "qcom,sa8775p-pcie-ep", .data = &cfg_1_34_0},
 	{ .compatible = "qcom,sdx55-pcie-ep", },
 	{ .compatible = "qcom,sm8450-pcie-ep", },
diff --git a/drivers/pci/endpoint/functions/pci-epf-mhi.c b/drivers/pci/endpoint/functions/pci-epf-mhi.c
index 7cc582c7e093..4dcc69f8e2d7 100644
--- a/drivers/pci/endpoint/functions/pci-epf-mhi.c
+++ b/drivers/pci/endpoint/functions/pci-epf-mhi.c
@@ -140,6 +140,23 @@ static const struct pci_epf_mhi_ep_info sa8775p_info = {
 	.flags = MHI_EPF_USE_DMA,
 };
 
+static struct pci_epf_header sa8797p_header = {
+	.vendorid = PCI_VENDOR_ID_QCOM,
+	.deviceid = 0x0120,
+	.baseclass_code = PCI_CLASS_OTHERS,
+	.interrupt_pin = PCI_INTERRUPT_INTA,
+};
+
+static const struct pci_epf_mhi_ep_info sa8797p_info = {
+	.config = &mhi_v1_config,
+	.epf_header = &sa8797p_header,
+	.bar_num = BAR_1,
+	.epf_flags = PCI_BASE_ADDRESS_MEM_TYPE_32,
+	.msi_count = 32,
+	.mru = 0x8000,
+	.flags = MHI_EPF_USE_DMA,
+};
+
 struct pci_epf_mhi {
 	const struct pci_epc_features *epc_features;
 	const struct pci_epf_mhi_ep_info *info;
@@ -936,6 +953,7 @@ static int pci_epf_mhi_probe(struct pci_epf *epf,
 }
 
 static const struct pci_epf_device_id pci_epf_mhi_ids[] = {
+	{ .name = "pci_epf_mhi_sa8797p", .driver_data = (kernel_ulong_t)&sa8797p_info },
 	{ .name = "pci_epf_mhi_sa8775p", .driver_data = (kernel_ulong_t)&sa8775p_info },
 	{ .name = "pci_epf_mhi_sdx55", .driver_data = (kernel_ulong_t)&sdx55_info },
 	{ .name = "pci_epf_mhi_sm8450", .driver_data = (kernel_ulong_t)&sm8450_info },
--
